src/lib/SaferNames/Inference.hs

```diff
@@ -1490,8 +1536,306 @@ freshType k = Var <$> freshInferenceName k
 freshEff :: Solver m => m n (EffectRow n)
 freshEff = EffectRow mempty . Just <$> freshInferenceName EffKind
 
-renameForPrinting :: (Type n, Type n) -> ((Type n, Type n), [AtomName n])
-renameForPrinting (t1, t2) = ((t1, t2), []) -- TODO!
+renameForPrinting :: (BindingsReader m, HoistableE e, InjectableE e, SubstE Name e)
+                  => e n -> m n (Abs (Nest (NameBinder AtomNameC)) e n)
+renameForPrinting e = do
+  WithBindings bs e' <- addBindings e
+  injectM $ renameForPrinting' bs e'
+
+renameForPrinting' :: (Distinct n, HoistableE e, InjectableE e, SubstE Name e)
+                   => Bindings n -> e n -> Abs (Nest (NameBinder AtomNameC)) e n
+renameForPrinting' bindings e = do
+  let infVars = filter (isInferenceVar bindings) $ freeVarsList AtomNameRep e
+  let ab = abstractFreeVars infVars e
+  let scope = toScope bindings
+  let hints = take (length infVars) $ map fromString $
+                map (:[]) ['a'..'z'] ++ map show [(0::Int)..]
+  withManyFresh hints AtomNameRep scope \bs' ->
+    runScopeReaderM (scope `extendOutMap` toScopeFrag bs') do
+      ab' <- injectM ab
+      e' <- applyNaryAbs ab' $ nestToList (inject . nameBinderName) bs'
+      return $ Abs bs' e'
+
+-- === dictionary synthesis ===
+
+-- main entrypoint to dictionary synthesizer
+trySynthDict :: (Emits n, Builder m, Fallible1 m, Scopable m, BindingsReader m)
+             => Type n -> m n (Atom n)
+trySynthDict ty = trySynthDictBlock ty >>= emitBlock
+
+trySynthDictBlock :: (Fallible1 m, BindingsReader m) => Type n -> m n (Block n)
+trySynthDictBlock ty = do
+  WithBindings bindings ty' <- addBindings ty
+  if hasInferenceVars bindings ty'
+    then
+      throw TypeErr "Can't synthesize a dictionary for a type with inference vars"
+    else do
+      let solutions = runSyntherM bindings $ buildBlock do
+                        ty'' <- injectM ty'
+                        synthDict ty''
+      let solutionsInstance = filter ((== UsedInstance) . snd) solutions
+      unless (length solutionsInstance <= 1) $
+        throw TypeErr $ "Multiple candidate class dictionaries for: " ++ pprint ty
+      case solutions of
+        [] -> throw TypeErr $ "Couldn't synthesize a class dictionary for: " ++ pprint ty
+        (d, _):_ -> injectM d
+
+-- TODO: we'd rather have something like this:
+--   data Givens n = Givens (M.Set (Type n) (Given n))
+-- but we need an Ord or Hashable instance on types
+data Givens n = Givens [Type n] [Given n]
```

Comment:
Are those two lists expected to be of equal length? Can't we get the type from the block type?

Reply:
Yes, the type is derivable from the Given type. This is a cache. But I forgot that `Given` is just `Block`, which has its own cached type, so I should have just used that. But I want to change it to a hash map soon so I'll leave it as-is for now.